guide.md

```diff
@@ -0,0 +1,440 @@
+users off-guard. For example, you can remove `.bazelrc` and rerun `bazel run`,
+and everything will work just fine: Bazel will simply fall back to
+`@local_config_cc`, because it's perfectly happy to use the C++ compiler
+executables that are still on the PATH.
+
+To mitigate this, we can use Nix to create a hermetic shell environment. This
+will provide the exact dependencies that developers need, and nothing else. As
+alluded to before, this is a similar concept to Python virtualenvs and other
+similar language-specific tools, except with unified support for system
+dependencies and language-specific dependencies across many different languages.
+
+The `nix-shell` command we used to get a Nix-provided Bazel was the simplest
+example of a Nix developer environment (one that simply added Bazel to the
+environment), but they can be much more sophisticated. We can create development
+environments that all contributors on a project can use with zero setup. When in
+the shell, the user has access to consistent versions of libraries for whichever
+languages are used in the project, as well as for system dependencies.
+
+To get started, put the following code in a file named `flake.nix` at the root
+of the repository:
+
+```nix
+{
+
+  inputs = {
+    nixpkgs.url = "github:NixOS/nixpkgs/nixos-22.05";
+    flake-utils.url = "github:numtide/flake-utils";
+  };
+
+  outputs = { nixpkgs, flake-utils, ... }:
+    # For every platform that Nix supports, we ...
+    flake-utils.lib.eachDefaultSystem (system:
+      # ... get the package set for this particular platform ...
+      let pkgs = import nixpkgs { inherit system; };
+      in {
+        # ... and define a development shell for it ...
+        devShells.default =
+          # ... with no globally-available CC toolchain ...
+          pkgs.mkShellNoCC {
+            name = "rules_nixpkgs_shell";
+            # ... which makes available the following dependencies, 
+            # all sourced from the `pkgs` package set:
+            packages = with pkgs; [ bazel_5 bazel-buildtools cacert nix git ];
+          };
+      });
+
+}
+```
+
+(For more about the `flake.nix` format, [this][flake-post] series of articles on
+the Tweag blog may be helpful.)
+
+We've also removed the "default" system-wide CC toolchain. The usual way to
+define a development shell is with the commonly-used `mkShell` Nix function,
+which creates a development shell with a given set of available dependencies
+_and_ a default CC toolchain. We've replaced it with `mkShellNoCC`, which, as
+the name suggests, does the same job but without implicitly making a CC
+toolchain available. This means it's no longer possible to unintentionally use
+`@local_config_cc`, as that toolchain won't be able to find the global binaries
+it wants and will fail to work.
+
+Now developers on different platforms can enter a development environment by
+running `nix develop`, and this environment will be identical across platforms.
```

Comment:
> this environment will be identical across platforms

Nit: I think I know what you're getting at, but it's a bit ambiguous. Is the user-guide only targeting Linux or is it also expected to work on MacOS? The reason I'm asking: Arguably "across platforms" would also mean across Linux and MacOS. But, in that case the environment is not identical, on MacOS it'll include Darwin executables instead of Linux ones.

Reply:
I've changed this to "consistent" instead of "identical". 

As for the first question: I wrote the guide with Linux in mind, but is there a reason why the guide wouldn't work on macOS?

Reply:
I've changed this to "consistent" instead of "identical". 

As for the first question: I wrote the guide with Linux in mind. Not sure if it would work elsewhere, but perhaps we can investigate macOS compatibility and try to make it work if we have a reasonable number of macOS users.

Reply:
Makes sense. Focusing on Linux first seems perfectly fair and I'm happy to leave MacOS for follow-up work. In that case it should just be stated somewhere that this guide is only targeting Linux and we'd also need an issue on rules_nixpkgs to track the MacOS support.